File: lambda/downloader.py
#!/usr/bin/env python3

import datetime
import json
import logging
import os
import traceback

import boto3
import requests
from botocore.exceptions import ClientError
from requests.exceptions import RequestException

from aws_xray_sdk.core import xray_recorder
from aws_xray_sdk.core import patch_all as xray_patch_all


logging.basicConfig()
_LOGGER = logging.getLogger("jmupark.lambda")
_LOGGER.setLevel(logging.DEBUG)

xray_patch_all()
s3 = boto3.client("s3")

_TIME_FORMAT = "%Y-%m-%d-%H-%M-%S"
_DOWNLOAD_URL = "https://www.jmu.edu/cgi-bin/parking_sign_data.cgi"
_URL_PARAMS = {
    "hash": "53616c7465645f5f2b3bed187af94815467e891744e87dc72a8d6f10cea4270a7fe8e2a08a3981a7a4bfb2ff402c4fdf8dc3cf0ae8b30a50f6ba85c7169b73b1edbafb488b046a19eb9588148d4e725f|869835tg89dhkdnbnsv5sg5wg0vmcf4mfcfc2qwm5968unmeh5"
}

BUCKET_NAME = os.getenv("BUCKET_NAME")


def lambda_handler(event, context):
    start_time = datetime.datetime.now()
    key = f"{start_time.strftime(_TIME_FORMAT)}.xml"

    try:
        data = requests.get(_DOWNLOAD_URL, params=_URL_PARAMS, timeout=7).content
        s3_response = s3.put_object(Bucket=BUCKET_NAME, Key=key, Body=data)

        response = {
            "statusCode": s3_response["ResponseMetadata"]["HTTPStatusCode"],
            "file": f"s3://{BUCKET_NAME}/{key}",
            "body": data,
            "size": len(data),
        }
    except (ClientError, RequestException, TypeError) as err:
        response = {
            "statusCode": 500,
            "error": str(err),
            "trace": "".join(traceback.format_exception(err))
        }

    _LOGGER.info("%s", json.dumps(response, default=str))
    return response


if __name__ == "__main__":
    lambda_handler(None, None)

File: lambda/downloader.ts
import axios from 'axios';
import { format } from 'date-fns';
import { S3 } from '@aws-sdk/client-s3';
import { Tracer } from '@aws-lambda-powertools/tracer';
import { Logger } from '@aws-lambda-powertools/logger';
import { ScheduledEvent, Context } from 'aws-lambda';

const tracer = new Tracer({ serviceName: 'jmuParkingDownloader' });
const logger = new Logger({ serviceName: 'jmuParkingDownloader' });
const s3 = tracer.captureAWSv3Client(new S3({ useFipsEndpoint: true }));

const BUCKET = process.env.BUCKET_NAME!;
const PARKING_URL = 'https://www.jmu.edu/cgi-bin/parking_sign_data.cgi';
const REQUIRED_URL_PARMS = {
  hash: '53616c7465645f5f2b3bed187af94815467e891744e87dc72a8d6f10cea4270a7fe8e2a08a3981a7a4bfb2ff402c4fdf8dc3cf0ae8b30a50f6ba85c7169b73b1edbafb488b046a19eb9588148d4e725f|869835tg89dhkdnbnsv5sg5wg0vmcf4mfcfc2qwm5968unmeh5',
};

// This is a separate function purely so that it can be tested
export function formatLexicographicTimestamp(date: Date): string {
  // format string should be equivalent to "%Y-%m-%d-%H-%M-%S"
  return format(date, 'yyyy-MM-dd-HH-mm-ss');
}

export async function handler(event: ScheduledEvent, context: Context): Promise<void> {
  logger.addContext(context);
  const startTime = new Date(event.time);
  const objectKey = `${formatLexicographicTimestamp(startTime)}.xml`;

  try {
    // Download data
    const response = await axios.get(PARKING_URL, { params: REQUIRED_URL_PARMS, transformResponse: (r) => r });
    const data: string = response.data;

    // Push to S3
    const s3Response = await s3.putObject({ Bucket: BUCKET, Key: objectKey, Body: data });
    logger.info('Data was successfully written to S3', {
      data: {
        statusCode: s3Response.$metadata.httpStatusCode,
        file: `s3://${BUCKET}/${objectKey}`,
        body: data,
        size: data.length,
      }
    });
  } catch (err) {
    logger.error('Error fetching or writing data', err as Error);
  }
}
